driver-core/src/main/com/mongodb/client/model/Indexes.java

```diff
@@ -0,0 +1,215 @@
+     * 2dsphere indexes allow field reordering; geoHaystack indexes require the first field to be the location field. Also, geoHaystack
+     * indexes are only usable via commands and so always return all results at once..
+     * </p>
+     *
+     * @param fieldName the field to create a geoHaystack index on
+     * @param additional the additional field that forms the geoHaystack index key
+     * @return the index specification
+     * @mongodb.driver.manual core/geohaystack geoHaystack index
+     */
+    public static Bson geoHaystack(final String fieldName, final Bson additional) {
+        notNull("fieldName", fieldName);
+        return compoundIndex(new BsonDocument(fieldName, new BsonString("geoHaystack")), additional);
+    }
+
+    /**
+     * Create an index key for a text index on the given field.
+     *
+     * @param fieldName the field to create a text index on
+     * @return the index specification
+     * @mongodb.driver.manual core/text text index
+     */
+    public static Bson text(final String fieldName) {
+        notNull("fieldName", fieldName);
+        return new BsonDocument(fieldName, new BsonString("text"));
+    }
+
+    /**
+     * Create an index key for a hashed index on the given field.
+     *
+     * @param fieldName the field to create a hashed index on
+     * @return the index specification
+     * @mongodb.driver.manual core/hashed hashed index
+     */
+    public static Bson hashed(final String fieldName) {
+        notNull("fieldName", fieldName);
+        return new BsonDocument(fieldName, new BsonString("hashed"));
+    }
+
+    /**
+     * create a compound index specifications.  If any field names are repeated, the last one takes precedence.
+     *
+     * @param indexes the index specifications
+     * @return the compound index specification
+     * @mongodb.driver.manual core/index-compound compoundIndex
+     */
+    public static Bson compoundIndex(final Bson... indexes) {
+        return compoundIndex(asList(indexes));
+    }
+
+    /**
+     * compound multiple index specifications.  If any field names are repeated, the last one takes precedence.
+     *
+     * @param indexes the index specifications
+     * @return the compound index specification
+     * @mongodb.driver.manual core/index-compound compoundIndex
+     */
+    public static Bson compoundIndex(final List<Bson> indexes) {
+        notNull("indexes", indexes);
+        return new Bson() {
+            @Override
+            public <TDocument> BsonDocument toBsonDocument(final Class<TDocument> documentClass, final CodecRegistry codecRegistry) {
+                BsonDocument compoundIndex = new BsonDocument();
+                for (Bson index : indexes) {
+                    BsonDocument indexDocument = index.toBsonDocument(documentClass, codecRegistry);
+                    for (String key : indexDocument.keySet()) {
+                        compoundIndex.remove(key);
```

Comment:
I see that Sorts class does this too, but now I don't see why the call to remove is necessary, as append will replace any existing mapping.